File: typescript/src/__tests__/dpop.test.ts
import { createHash, createPublicKey, verify } from 'crypto';
import { generateAgentKey, MemoryKeyStore } from '../keys';
import { mintDPoPProof, mintDPoPProofWithKeyStore } from '../dpop';

const CHECK_URL = 'https://api.agenttrust.id/api/v1/agenttrust/check';

function parseProof(proof: string) {
  const [h, p, s] = proof.split('.');
  return {
    header: JSON.parse(Buffer.from(h, 'base64url').toString()),
    payload: JSON.parse(Buffer.from(p, 'base64url').toString()),
    signingInput: `${h}.${p}`,
    sig: Buffer.from(s, 'base64url'),
  };
}

function verifyProof(proof: string): boolean {
  const { header, signingInput, sig } = parseProof(proof);
  const pub = createPublicKey({ key: header.jwk, format: 'jwk' });
  return verify(null, Buffer.from(signingInput), pub, sig);
}

describe('mintDPoPProof', () => {
  it('produces a verifiable proof with the platform-expected claims', () => {
    const kp = generateAgentKey();
    const proof = mintDPoPProof(kp.privateKeyPem, 'POST', CHECK_URL, 'access-tok');
    const { header, payload } = parseProof(proof);

    expect(header.typ).toBe('dpop+jwt');
    expect(header.alg).toBe('EdDSA');
    expect(header.jwk.kty).toBe('OKP');
    expect(header.jwk.crv).toBe('Ed25519');
    expect(header.jwk.x).toBeTruthy();

    expect(payload.htm).toBe('POST');
    expect(payload.htu).toBe(CHECK_URL);
    expect(typeof payload.iat).toBe('number');
    expect(payload.jti).toBeTruthy();
    expect(payload.ath).toBe(createHash('sha256').update('access-tok').digest().toString('base64url'));

    expect(verifyProof(proof)).toBe(true);
  });

  it('omits ath when no access token is bound', () => {
    const kp = generateAgentKey();
    const { payload } = parseProof(mintDPoPProof(kp.privateKeyPem, 'GET', 'https://x/y', ''));
    expect(payload.ath).toBeUndefined();
  });

  it('mints unique jti per call', () => {
    const kp = generateAgentKey();
    const a = parseProof(mintDPoPProof(kp.privateKeyPem, 'POST', CHECK_URL, 'tok'));
    const b = parseProof(mintDPoPProof(kp.privateKeyPem, 'POST', CHECK_URL, 'tok'));
    expect(a.payload.jti).not.toBe(b.payload.jti);
  });

  it('rejects a non-Ed25519 key path gracefully', () => {
    expect(() => mintDPoPProof('-----BEGIN PRIVATE KEY-----\nbogus\n-----END PRIVATE KEY-----', 'POST', CHECK_URL, '')).toThrow();
  });
});

describe('mintDPoPProofWithKeyStore', () => {
  it('signs through the KeyStore and verifies against the embedded jwk', () => {
    const kp = generateAgentKey();
    const ks = new MemoryKeyStore();
    ks.store('agent-1', kp.privateKeyPem);

    const proof = mintDPoPProofWithKeyStore(ks, 'agent-1', kp.publicKeyPem, 'POST', CHECK_URL, 'tok');
    expect(verifyProof(proof)).toBe(true);
  });
});

File: go/dpop_test.go
package agenttrust

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"strings"
	"testing"
)

// parseDPoP splits a compact DPoP JWT and returns its decoded header, payload,
// the signing input, and the raw signature — enough to validate structure and
// verify the signature against the embedded JWK.
func parseDPoP(t *testing.T, proof string) (header, payload map[string]interface{}, signingInput string, sig []byte) {
	t.Helper()
	parts := strings.Split(proof, ".")
	if len(parts) != 3 {
		t.Fatalf("expected 3 JWT parts, got %d", len(parts))
	}
	hb, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		t.Fatalf("decode header: %v", err)
	}
	pb, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		t.Fatalf("decode payload: %v", err)
	}
	sig, err = base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		t.Fatalf("decode sig: %v", err)
	}
	if err := json.Unmarshal(hb, &header); err != nil {
		t.Fatalf("unmarshal header: %v", err)
	}
	if err := json.Unmarshal(pb, &payload); err != nil {
		t.Fatalf("unmarshal payload: %v", err)
	}
	return header, payload, parts[0] + "." + parts[1], sig
}

func TestMintDPoPProof_StructureAndSignature(t *testing.T) {
	kp, err := GenerateAgentKey()
	if err != nil {
		t.Fatalf("keygen: %v", err)
	}
	const tok = "wimse.token.value"
	proof, err := MintDPoPProof(kp.PrivateKeyPEM, "POST", "https://api.agenttrust.id/api/v1/agenttrust/check", tok)
	if err != nil {
		t.Fatalf("mint: %v", err)
	}

	header, payload, signingInput, sig := parseDPoP(t, proof)

	if header["typ"] != "dpop+jwt" {
		t.Errorf("typ = %v, want dpop+jwt", header["typ"])
	}
	if header["alg"] != "EdDSA" {
		t.Errorf("alg = %v, want EdDSA", header["alg"])
	}
	jwk, ok := header["jwk"].(map[string]interface{})
	if !ok || jwk["kty"] != "OKP" || jwk["crv"] != "Ed25519" {
		t.Fatalf("jwk = %v, want OKP/Ed25519", header["jwk"])
	}
	if payload["htm"] != "POST" {
		t.Errorf("htm = %v, want POST", payload["htm"])
	}
	if payload["htu"] != "https://api.agenttrust.id/api/v1/agenttrust/check" {
		t.Errorf("htu = %v", payload["htu"])
	}
	if payload["jti"] == nil || payload["jti"] == "" {
		t.Error("jti missing")
	}

	// ath must be base64url(sha256(token)).
	sum := sha256.Sum256([]byte(tok))
	if payload["ath"] != base64.RawURLEncoding.EncodeToString(sum[:]) {
		t.Errorf("ath = %v, want hash of token", payload["ath"])
	}

	// Signature must verify against the public key embedded in the jwk.
	x, err := base64.RawURLEncoding.DecodeString(jwk["x"].(string))
	if err != nil || len(x) != ed25519.PublicKeySize {
		t.Fatalf("bad jwk.x: %v", err)
	}
	if !ed25519.Verify(ed25519.PublicKey(x), []byte(signingInput), sig) {
		t.Error("DPoP signature did not verify against embedded jwk")
	}
}

func TestMintDPoPProof_NoAccessToken_OmitsATH(t *testing.T) {
	kp, err := GenerateAgentKey()
	if err != nil {
		t.Fatalf("keygen: %v", err)
	}
	proof, err := MintDPoPProof(kp.PrivateKeyPEM, "GET", "https://api.agenttrust.id/v1/agents", "")
	if err != nil {
		t.Fatalf("mint: %v", err)
	}
	_, payload, _, _ := parseDPoP(t, proof)
	if _, present := payload["ath"]; present {
		t.Error("ath should be omitted when no access token is provided")
	}
}

func TestMintDPoPProof_InvalidKeyRejected(t *testing.T) {
	if _, err := MintDPoPProof("not-a-pem", "GET", "https://x/y", ""); err == nil {
		t.Fatal("expected error for invalid private key PEM")
	}
}

func TestMintDPoPProofWithKeyStore_VerifiesAndOmitsExportedKey(t *testing.T) {
	kp, err := GenerateAgentKey()
	if err != nil {
		t.Fatalf("keygen: %v", err)
	}
	ks := NewMemoryKeyStore()
	if err := ks.Store("agent-1", kp.PrivateKeyPEM); err != nil {
		t.Fatalf("store: %v", err)
	}

	// Signs through the KeyStore; only the public key PEM is supplied.
	proof, err := MintDPoPProofWithKeyStore(ks, "agent-1", kp.PublicKeyPEM,
		"POST", "https://api.agenttrust.id/api/v1/agenttrust/check", "tok")
	if err != nil {
		t.Fatalf("mint: %v", err)
	}

	header, payload, signingInput, sig := parseDPoP(t, proof)
	if header["typ"] != "dpop+jwt" || header["alg"] != "EdDSA" {
		t.Fatalf("bad header: %v", header)
	}
	jwk := header["jwk"].(map[string]interface{})
	x, err := base64.RawURLEncoding.DecodeString(jwk["x"].(string))
	if err != nil || len(x) != ed25519.PublicKeySize {
		t.Fatalf("bad jwk.x: %v", err)
	}
	if !ed25519.Verify(ed25519.PublicKey(x), []byte(signingInput), sig) {
		t.Error("KeyStore-minted DPoP did not verify against embedded jwk")
	}
	if payload["jti"] == nil || payload["jti"] == "" {
		t.Error("jti missing")
	}
}

func TestMintDPoPProofWithKeyStore_NilKeyStore(t *testing.T) {
	if _, err := MintDPoPProofWithKeyStore(nil, "a", "", "GET", "https://x/y", ""); err == nil {
		t.Fatal("expected error for nil KeyStore")
	}
}
